### PR TITLE
8309673: Refactor ref_at methods in SA ConstantPool

### PR DESCRIPTION
**Pull request by matias9927**
The accessor methods in constantpool.cpp were previously cleaned up to allow for different types of indices to be used, distinguishing them by the bytecode. This patch adds the same changes to the hotspot serviceability agent code. Verified with tier 1-5 tests.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8309673](https://bugs.openjdk.org/browse/JDK-8309673): Refactor ref_at methods in SA ConstantPool (**Enhancement** - P4)

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [2ee60bd1](https://git.openjdk.org/jdk/pull/14385/files/2ee60bd190dd89ea836b7bb207bb183168ec1d72)
 * [Frederic Parain](https://openjdk.org/census#fparain) (@fparain - **Reviewer**) ⚠️ Review applies to [2ee60bd1](https://git.openjdk.org/jdk/pull/14385/files/2ee60bd190dd89ea836b7bb207bb183168ec1d72)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14385/head:pull/14385` \
`$ git checkout pull/14385`

Update a local copy of the PR: \
`$ git checkout pull/14385` \
`$ git pull https://git.openjdk.org/jdk.git pull/14385/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14385`

View PR using the GUI difftool: \
`$ git pr show -t 14385`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14385.diff">https://git.openjdk.org/jdk/pull/14385.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14385#issuecomment-1583447507)

### CODE REVIEW

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/oops/ConstantPool.java [hunk 1]
```diff
@@ -252,6 +253,31 @@ public int getFieldOrMethodAt(int which) {
     return res;
   }
 
+  // Translate index, which could be CPCache index or Indy index, to a constant pool index
+  public int to_cp_index(int index, int code) {
+    Assert.that(getCache() != null, "'index' is a rewritten index so this class must have been rewritten");
+    switch(code) {
+      case Bytecodes._invokedynamic:
+        int poolIndex = getCache().getIndyEntryAt(index).getConstantPoolIndex();
+        return poolIndex = invokeDynamicNameAndTypeRefIndexAt(poolIndex);
```
Comment (coleenp) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/oops/ConstantPool.java:262]: probably don't need another assignment to poolIndex in the return statement.

**Review by coleenp — APPROVED** [review]
Looks good.

**Review by fparain — APPROVED** [review]

**Review by iklam — APPROVED** [review]
LGTM. I would suggest changing "Serviceability" in the title to "SA"

**Comment by matias9927** [comment]
Thank you for the reviews @coleenp, @fparain, and @iklam !
/integrate

**Pull request closed without merging**